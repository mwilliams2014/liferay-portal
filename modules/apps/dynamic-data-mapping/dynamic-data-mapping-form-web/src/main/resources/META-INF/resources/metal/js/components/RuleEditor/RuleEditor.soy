{namespace RuleEditor}

/**
 * Prints the DDM rule settings.
 */
{template .render}
	{@param conditions: list<[
			operator: string,
			operands: list<[
				type: string,
				label: string,
				value: string
			]>
		]>
	}
	{@param firstOperandList : list<[
			value: string,
			type: string,
			name: string,
			options: list<[
				label: string,
				value: string
			]>
		]>
	}
	{@param logicalOperator: string}
	{@param readOnly: bool}
	{@param strings: ?}
	{@param? _handleFirstOperandSelection: any}
	{@param? _handleOperatorSelection: any}
	{@param? _handleSecondOperandSelection: any}
	{@param? _handleTypeSelection: any}
	{@param? operators:  list<[
			value: string,
			name: string,
			parameterTypes: list<string>,
			returnType: string
		]>
	}
	{@param? secondOperandTypeList: list<[
			value: string,
			name: string
		]>
	}
	{@param? secondOperandTypeSelectedList: list<[
			value: string,
			name: string
		]>
	}
	{@param? spritemap: string}

	<div class="form-builder-rule-builder">
		<h2 class="form-builder-section-title text-default">{$strings.title}</h2>
		<h4 class="text-default">{$strings.description}</h4>

		<ul class="liferay-ddm-form-builder-rule-condition-list liferay-ddm-form-rule-builder-timeline timeline {length($conditions) > 1 ? 'can-remove-item' : ''}">
			{let $enableLogicalOperator: (length($conditions) > 1) ? true : false /}

			{call .rulesHeader}
				{param extraContent kind="html"}
					{call .logicalOperatorDropDown}
						{param logicalOperator: $strings[$logicalOperator] /}
						{param strings: $strings /}
						{param enableLogicalOperator: $enableLogicalOperator /}
					{/call}
				{/param}
				{param logicalOperator: $strings[$logicalOperator] /}
				{param title: $strings.condition /}
			{/call}

			{foreach $condition in $conditions}
				{call .condition}
					{param _handleFirstOperandSelection: $_handleFirstOperandSelection /}
					{param _handleOperatorSelection: $_handleOperatorSelection /}
					{param _handleSecondOperandSelection: $_handleSecondOperandSelection /}
					{param _handleTypeSelection: $_handleTypeSelection /}
					{param firstOperandList: $firstOperandList /}
					{param firstOperandValueSelected: $condition.operands[0].value /}
					{param if: $strings.if /}
					{param index: index($condition) /}
					{param logicalOperator: $strings[$logicalOperator] /}
					{param operators: $operators /}
					{param operatorSelected: $condition.operator /}
					{param readOnly: $readOnly /}
					{param secondOperandTypeList: $secondOperandTypeList /}
					{param secondOperandTypeName: '' /}
					{param secondOperandTypeSelected: $condition.operands[0].type /}
					{param secondOperandTypeSelectedList: $secondOperandTypeSelectedList /}
					{param secondOperandTypeValue: isNonnull($condition.operands[1]) ? $condition.operands[1].value : '' /}
					{param spritemap: $spritemap /}
				{/call}

				{ifempty}
					{call .condition}
						{param _handleFirstOperandSelection: $_handleFirstOperandSelection /}
						{param _handleOperatorSelection: $_handleOperatorSelection /}
						{param _handleSecondOperandSelection: $_handleSecondOperandSelection /}
						{param _handleTypeSelection: $_handleTypeSelection /}
						{param firstOperandList: $firstOperandList /}
						{param if: $strings.if /}
						{param index: 0 /}
						{param logicalOperator: $strings[$logicalOperator] /}
						{param operators: $operators /}
						{param readOnly: $readOnly /}
						{param secondOperandTypeList: $secondOperandTypeList /}
						{param secondOperandTypeName: '' /}
						{param spritemap: $spritemap /}
					{/call}
			{/foreach}
		</ul>
	</div>
{/template}

{template .condition}
	{@param firstOperandList : list<[
			value: string,
			type: string,
			name: string,
			options: list<[
				label: string,
				value: string
			]>
		]>
	}
	{@param if: string}
	{@param index: int}
	{@param readOnly: bool}
	{@param? _handleFirstOperandSelection: any}
	{@param? _handleOperatorSelection: any}
	{@param? _handleSecondOperandSelection: any}
	{@param? _handleTypeSelection: any}
	{@param? firstOperandValueSelected: string}
	{@param? operators:  list<[
			value: string,
			name: string,
			parameterTypes: list<string>,
			returnType: string
		]>
	}
	{@param? operatorSelected: string}
	{@param? secondOperandTypeName: string }
	{@param? secondOperandTypeList : list<[
			value: string,
			name: string
		]>
	}
	{@param? secondOperandTypeSelectedList : list<[
			value: string,
			name: string
		]>
	}
	{@param? secondOperandTypeValue: any }
	{@param? spritemap: string}

	<li class="form-builder-rule-condition-container timeline-item" condition-index="{$index}">
		<div class="panel panel-default">
			<div class="flex-container panel-body">
				<div class="form-group-autofit">
					<div class="form-group-item form-group-item-label form-group-item-shrink">
						<h4>
							<span class="text-truncate-inline">
								<span class="text-truncate">{$if}</span>
							</span>
						</h4>
					</div>

					<div class="condition-if form-group-item" condition-if-index="{$index}">
						{delcall PageRenderer.RegisterFieldType variant="'select'"}
							{param events: [
								'fieldEdited': $_handleFirstOperandSelection
							] /}
							{param options: $firstOperandList /}
							{param ref: 'conditionIf' /}
							{param spritemap: $spritemap /}
							{param value: [$firstOperandValueSelected]   /}
						{/delcall}
					</div>

					{call .operators}
						{param _handleOperatorSelection: $_handleOperatorSelection /}
						{param index: $index /}
						{param operators: $operators /}
						{param operatorSelected: $operatorSelected /}
						{param spritemap: $spritemap /}
					{/call}

					{call .secondOperandTypeList}
						{param _handleTypeSelection: $_handleTypeSelection /}
						{param index: $index /}
						{param operatorSelected: $operatorSelected /}
						{param secondOperandTypeList: $secondOperandTypeList /}
						{param secondOperandTypeSelectedList: $secondOperandTypeSelectedList /}
						{param spritemap: $spritemap /}
					{/call}

					{if isNonnull($secondOperandTypeSelectedList) and isNonnull($secondOperandTypeSelectedList[$index])}
						{let $textFieldVisible: ($secondOperandTypeSelectedList[$index].name == 'text') ? true : false /}

						<div class="condition-type-value form-group-item {if not $textFieldVisible}hide{/if}" condition-type-value-index="{$index}">
							{delcall PageRenderer.RegisterFieldType variant="'text'"}
								{param events: [
									'fieldEdited': $_handleSecondOperandSelection
								] /}
								{param name: $secondOperandTypeName /}
								{param readOnly: $readOnly /}
								{param ref: 'typeValueInput' /}
								{param spritemap: $spritemap /}
								{param value: $secondOperandTypeValue /}
							{/delcall}
						</div>

						{let $selectFieldVisible: ($secondOperandTypeSelectedList[$index].name == 'field') ? true : false /}

						<div class="condition-type-value-select form-group-item {if not $selectFieldVisible}hide{/if}" condition-type-value-select-index="{$index}">
							{delcall PageRenderer.RegisterFieldType variant="'select'"}
								{param events: [
										'fieldEdited': $_handleSecondOperandSelection
									] /}
								{param options: $firstOperandList /}
								{param ref: 'typeValueSelect' /}
								{param spritemap: $spritemap /}
								{param value: [$secondOperandTypeValue] /}
							{/delcall}
						</div>

						{let $selectOptionsFieldVisible: ($secondOperandTypeSelectedList[$index].name == 'select') ? true : false /}

						{foreach $operand in $firstOperandList}
							{if $operand.value == $firstOperandValueSelected}
								<div class="condition-type-value-select-options form-group-item {if not $selectOptionsFieldVisible}hide{/if}" condition-type-value-select-options-index="{$index}">
									{delcall PageRenderer.RegisterFieldType variant="'select'"}
										{param events: [
												'fieldEdited': $_handleSecondOperandSelection
											] /}
										{param options: $operand.options /}
										{param ref: 'typeValueSelectOptions' /}
										{param spritemap: $spritemap /}
										{param value: [$secondOperandTypeValue] /}
									{/delcall}
								</div>
							{/if}
						{/foreach}
					{/if}
				</div>

				<div class="timeline-increment">
					<span class="timeline-icon"></span>
				</div>
			</div>
		</div>
	</li>
{/template}

{template .operators}
	{@param index: int}
	{@param? _handleOperatorSelection: any}
	{@param? operators:  list<[
			value: string,
			name: string,
			parameterTypes: list<string>,
			returnType: string
		]>
	}
	{@param? operatorSelected: string}
	{@param? spritemap: string}

	{let $readOnly: ($operators and length($operators) > 0) ? false : true /}

	<div class="condition-operator form-group-item" condition-operator-index="{$index}">
		{delcall PageRenderer.RegisterFieldType variant="'select'"}
			{param events: [
				'fieldEdited': $_handleOperatorSelection
			] /}
			{param options: $operators /}
			{param readOnly: $readOnly /}
			{param ref: 'conditionOperator' /}
			{param spritemap: $spritemap /}
			{param value: [$operatorSelected] /}
		{/delcall}
	</div>
{/template}

{template .secondOperandTypeList}
	{@param index: int}
	{@param? _handleTypeSelection: any}
	{@param? operatorSelected: any}
	{@param? secondOperandTypeList : list<[
			value: string,
			name: string
		]>
	}
	{@param? secondOperandTypeSelectedList : list<[
			value: string,
			name: string
		]>
	}
	{@param? spritemap: string}

	{let $visible: $operatorSelected and isNonnull($secondOperandTypeSelectedList) and isNonnull($secondOperandTypeSelectedList[$index]) ? true : false /}

	<div class="condition-type form-group-item {if not $visible}hide{/if}" condition-type-index="{$index}">
		{delcall PageRenderer.RegisterFieldType variant="'select'"}
			{param events: [
				'fieldEdited': $_handleTypeSelection
			] /}
			{param options: $secondOperandTypeList /}
			{param ref: 'type' /}
			{param spritemap: $spritemap /}
			{param value: [$visible ? $secondOperandTypeSelectedList?[$index].value : ''] /}
		{/delcall}
	</div>
{/template}

/**
 * Prints Logical Operator Dropdown
 */
{template .logicalOperatorDropDown}
	{@param strings: ?}
	{@param? enableLogicalOperator: bool}
	{@param? logicalOperator: string}

	{let $attributes kind="attributes"}
		class="btn btn-default dropdown-toggle dropdown-toggle-operator text-uppercase"
		data-toggle="dropdown"
		type="button"

		{if not $enableLogicalOperator}
			disabled
		{/if}
	{/let}

	<div class="btn-group dropdown" style="block">
		<button {$attributes}>
			<span class="dropdown-toggle-selected-value">{$logicalOperator}</span> <span class="caret"></span>
		</button>

		<ul class="dropdown-menu">
			<li class="logic-operator text-uppercase" data-logical-operator-value="or"><a href="#">{$strings['or']}</a></li>
			<li class="divider"></li>
			<li class="logic-operator text-uppercase" data-logical-operator-value="and"><a href="#">{$strings['and']}</a></li>
		</ul>
	</div>
{/template}

/**
 * Prints rules header.
 */
{template .rulesHeader}
	{@param title: string}
	{@param? extraContent: html}

	<li class="timeline-item">
		<div class="panel panel-default">
			<div class="flex-container panel-body">
				<div class="h4 panel-title">
					{$title}
				</div>

				{if $extraContent}
					{$extraContent}
				{/if}

				<div class="timeline-increment">
					<span class="timeline-icon"></span>
				</div>
			</div>
		</div>
	</li>
{/template}