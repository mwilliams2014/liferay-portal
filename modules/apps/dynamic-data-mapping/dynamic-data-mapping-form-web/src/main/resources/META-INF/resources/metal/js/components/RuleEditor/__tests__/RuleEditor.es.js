import RuleEditor from '../RuleEditor.es';
import './__fixtures__/RuleEditorMockField.es';

let component;

const spritemap = 'icons.svg';

const pages = [
	{
		rows: [
			{
				columns: [
					{
						fields: [
							{
								fieldName: 'Nome',
								label: 'Nome',
								options: [
									{
										label: 'Option',
										value: 'Option'
									}
								],
								type: 'text'
							},
							{
								fieldName: 'Sobrenome',
								label: 'text',
								options: [
									{
										label: 'Option',
										value: 'Option'
									}
								],
								type: 'text'
							},
							{
								fieldName: 'EstadoCivil',
								label: 'Estado civil',
								options: [
									{
										label: 'Casado',
										value: 'Casado'
									},
									{
										label: 'Solteiro',
										value: 'Solteiro'
									}
								],
								type: 'select'
							}
						]
					}
				]
			}
		]
	}
];

const firstOperandList = [
	{
		dataType: 'string',
		name: 'Nome',
		options: [
			{
				label: 'Option',
				value: 'Option'
			}
		],
		type: 'text',
		value: 'Nome'
	},
	{
		dataType: 'string',
		name: 'text',
		options: [
			{
				label: 'Option',
				value: 'Option'
			}
		],
		type: 'text',
		value: 'Sobrenome'
	},
	{
		dataType: 'string',
		name: 'EstadoCivil',
		options: [
			{
				label: 'Casado',
				value: 'Casado'
			},
			{
				label: 'Solteiro',
				value: 'Solteiro'
			}
		],
		type: 'select',
		value: 'Estado civil'
	}
];

const secondOperandTypeSelectedList = [
	{
		name: '',
		value: ''
	}
];

const operatorsList = [
	{name: 'contains', parameterTypes: ['text', 'text'], returnType: 'boolean', value: 'Contains'},
	{name: 'equals-to', parameterTypes: ['text', 'text'], returnType: 'boolean', value: 'Is equal to'},
	{name: 'is-empty', parameterTypes: ['text'], returnType: 'boolean', value: 'Is empty'},
	{name: 'not-contains', parameterTypes: ['text', 'text'], returnType: 'boolean', value: 'Does not contain'},
	{name: 'not-equals-to', parameterTypes: ['text', 'text'], returnType: 'boolean', value: 'Is not equal to'},
	{name: 'not-is-empty', parameterTypes: ['text'], returnType: 'boolean', value: 'Is not empty'}
];

const functionsMetadata = {text: [
	{name: 'contains', parameterTypes: ['text', 'text'], returnType: 'boolean', value: 'Contains'},
	{name: 'equals-to', parameterTypes: ['text', 'text'], returnType: 'boolean', value: 'Is equal to'},
	{name: 'is-empty', parameterTypes: ['text'], returnType: 'boolean', value: 'Is empty'},
	{name: 'not-contains', parameterTypes: ['text', 'text'], returnType: 'boolean', value: 'Does not contain'},
	{name: 'not-equals-to', parameterTypes: ['text', 'text'], returnType: 'boolean', value: 'Is not equal to'},
	{name: 'not-is-empty', parameterTypes: ['text'], returnType: 'boolean', value: 'Is not empty'}
]};

describe(
	'RuleEditor',
	() => {
		beforeEach(
			() => {
				jest.useFakeTimers();
			}
		);
		afterEach(
			() => {
				if (component) {
					component.dispose();
				}
			}
		);

		it(
			'should hide third and fourth selectors by default',
			() => {
				const conditions = [];

				component = new RuleEditor(
					{
						conditions,
						functionsMetadata,
						pages,
						secondOperandTypeSelectedList,
						spritemap
					}
				);

				const type = document.querySelector('.condition-type').classList.contains('hide');
				const typeValue = document.querySelector('.condition-type-value');
				const typeValueSelect = document.querySelector('.condition-type-value-select');

				expect(type && !typeValue && !typeValueSelect).toEqual(true);
			}
		);

		it(
			'should show third selector when first and second selectors have options selected',
			() => {
				const conditions = [
					{
						operands: [
							{
								type: 'text',
								value: 'Nome'
							}
						],
						operator: 'Is not equal to'
					}
				];

				const secondOperandTypeList = [
					{
						name: 'text',
						value: 'Value'
					},
					{
						name: 'field',
						value: 'other-field'
					}
				];

				component = new RuleEditor(
					{
						conditions,
						firstOperandList,
						functionsMetadata,
						logicalOperator: 'or',
						operatorsList,
						pages,
						secondOperandTypeList,
						secondOperandTypeSelectedList,
						spritemap
					}
				);

				component.refs.conditionIf.emitFieldEdited('Nome', 'Nome');
				component.refs.conditionOperator.emitFieldEdited('Is not equal to', 'not-equals-to');

				const type = document.querySelector('.condition-type').classList.contains('hide');
				const typeValue = document.querySelector('.condition-type-value').classList.contains('hide');
				const typeValueSelect = document.querySelector('.condition-type-value-select').classList.contains('hide');

				expect(!type && typeValue && typeValueSelect).toEqual(true);
			}
		);

		it(
			'should reset and show fourth input selector when first, second and third have options selected and first selector is text field',
			() => {
				const conditions = [
					{
						operands: [
							{
								type: 'text',
								value: 'Nome'
							},
							{
								type: 'text',
								value: 'Nome'
							}
						],
						operator: 'Is not equal to'
					}
				];

				const secondOperandTypeList = [
					{
						name: 'text',
						value: 'Value'
					},
					{
						name: 'field',
						value: 'other-field'
					}
				];

				const secondOperandTypeSelectedList = [
					{
						name: 'text',
						value: 'Value'
					}
				];

				component = new RuleEditor(
					{
						conditions,
						firstOperandList,
						functionsMetadata,
						pages,
						secondOperandTypeList,
						secondOperandTypeSelectedList,
						spritemap
					}
				);

				component.refs.conditionIf.emitFieldEdited('', 'Nome');
				component.refs.conditionOperator.emitFieldEdited('Is not equal to', 'not-equals-to');
				component.refs.type.emitFieldEdited('Value', 'text');

				const secondOperandEmpty = (conditions[0].operands[1].type === '') && (conditions[0].operands[1].value === '');
				const type = document.querySelector('.condition-type').classList.contains('hide');
				const typeValue = document.querySelector('.condition-type-value').classList.contains('hide');
				const typeValueSelect = document.querySelector('.condition-type-value-select').classList.contains('hide');

				expect(
					!type && !typeValue &&
					typeValueSelect && secondOperandEmpty
				).toEqual(true);
			}
		);

		it(
			'should show fourth "select" selector when first, second and third have options selected and first selector is select field',
			() => {
				const conditions = [
					{
						operands: [
							{
								type: 'select',
								value: 'Estado civil'
							}
						],
						operator: 'Is not equal to'
					}
				];

				const secondOperandTypeSelectedList = [
					{
						name: 'select',
						value: 'Value'
					}
				];

				component = new RuleEditor(
					{
						conditions,
						firstOperandList,
						functionsMetadata,
						pages,
						secondOperandTypeSelectedList,
						spritemap
					}
				);

				component.refs.conditionIf.emitFieldEdited('', 'EstadoCivil');
				component.refs.conditionOperator.emitFieldEdited('Is not equal to', 'not-equals-to');
				component.refs.type.emitFieldEdited('Value', 'select');

				const type = document.querySelector('.condition-type').classList.contains('hide');
				const typeValue = document.querySelector('.condition-type-value').classList.contains('hide');
				const typeValueSelect = document.querySelector('.condition-type-value-select').classList.contains('hide');
				const typeValueSelectOptions = document.querySelector('.condition-type-value-select-options').classList.contains('hide');

				expect(
					!type && typeValue &&
					typeValueSelect && !typeValueSelectOptions
				).toEqual(true);
			}
		);

		it(
			'should show all fields when fourth field is a text input',
			() => {
				const conditions = [
					{
						operands: [
							{
								type: 'text',
								value: 'Nome'
							},
							{
								type: '',
								value: 'Any text'
							}
						],
						operator: 'Is not equal to'
					}
				];

				const secondOperandTypeSelectedList = [
					{
						name: 'text',
						value: 'Value'
					}
				];

				component = new RuleEditor(
					{
						conditions,
						firstOperandList,
						functionsMetadata,
						pages,
						secondOperandTypeSelectedList,
						spritemap
					}
				);

				component.refs.conditionIf.emitFieldEdited('', 'Nome');
				component.refs.conditionOperator.emitFieldEdited('Is not equal to', 'not-equals-to');
				component.refs.type.emitFieldEdited('Value', 'text');
				component.refs.typeValueInput.emitFieldEdited('', 'Nome');

				expect(component).toMatchSnapshot();
			}
		);

		it(
			'should show all fields when fourth field is a select field',
			() => {
				const conditions = [
					{
						operands: [
							{
								type: 'text',
								value: 'Nome'
							},
							{
								type: 'text',
								value: 'Nome'
							}
						],
						operator: 'Is not equal to'
					}
				];

				const secondOperandTypeSelectedList = [
					{
						name: 'field',
						value: 'other-field'
					}
				];

				component = new RuleEditor(
					{
						conditions,
						firstOperandList,
						functionsMetadata,
						pages,
						secondOperandTypeSelectedList,
						spritemap
					}
				);

				component.refs.conditionIf.emitFieldEdited('', 'Nome');
				component.refs.conditionOperator.emitFieldEdited('Is not equal to', 'not-equals-to');
				component.refs.type.emitFieldEdited('other-field', 'field');
				component.refs.typeValueSelect.emitFieldEdited('', 'Nome');

				expect(component).toMatchSnapshot();
			}
		);
	}
);

describe(
	'Regression',
	() => {
		it(
			'LPS-85642 Should hide second operand and value when operator is reset',
			() => {
				component = new RuleEditor(
					{
						functionsMetadata,
						pages,
						spritemap
					}
				);

				component.refs.conditionIf.emitFieldEdited('', 'Nome');
				component.refs.conditionOperator.emitFieldEdited('Is not equal to', 'not-equals-to');
				component.refs.type.emitFieldEdited('other-field', 'field');

				jest.runAllTimers();

				component.refs.typeValueSelect.emitFieldEdited('', 'Nome');
				component.refs.conditionOperator.emitFieldEdited('', '');

				jest.runAllTimers();

				expect(component).toMatchSnapshot();
			}
		);

		it(
			'LPS-85644 Should clear first operand when field is removed from pages',
			() => {
				component = new RuleEditor(
					{
						functionsMetadata,
						pages,
						spritemap
					}
				);

				component.refs.conditionIf.emitFieldEdited('Nome', 'Nome');
				component.refs.conditionOperator.emitFieldEdited('Is not equal to', 'not-equals-to');
				component.refs.type.emitFieldEdited('other-field', 'field');

				component.pages = [];

				jest.runAllTimers();

				expect(component).toMatchSnapshot();
			}
		);
	}
);