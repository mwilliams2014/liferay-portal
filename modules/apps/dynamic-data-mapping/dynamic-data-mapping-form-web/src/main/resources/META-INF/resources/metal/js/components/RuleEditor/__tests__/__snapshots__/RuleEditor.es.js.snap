// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Regression LPS-85642 Should hide second operand and value when operator is reset 1`] = `
<div class="form-builder-rule-builder">
  <h2 class="form-builder-section-title text-default">rule</h2>
  <h4 class="text-default">define-condition-and-action-to-change-fields-and-elements-on-the-form</h4>
  <ul class="liferay-ddm-form-builder-rule-condition-list liferay-ddm-form-rule-builder-timeline timeline ">
    <li class="timeline-item">
      <div class="panel panel-default">
        <div class="flex-container panel-body">
          <div class="h4 panel-title">condition</div>
          <div class="btn-group dropdown" style="">
            <button class="btn btn-default dropdown-toggle dropdown-toggle-operator text-uppercase" data-toggle="dropdown" type="button" disabled="">
              <span class="dropdown-toggle-selected-value">or</span>
              <span class="caret"></span>
            </button>
            <ul class="dropdown-menu">
              <li class="logic-operator text-uppercase" data-logical-operator-value="or">
                <a href="#">or</a>
              </li>
              <li class="divider"></li>
              <li class="logic-operator text-uppercase" data-logical-operator-value="and">
                <a href="#">and</a>
              </li>
            </ul>
          </div>
          <div class="timeline-increment">
            <span class="timeline-icon"></span>
          </div>
        </div>
      </div>
    </li>
    <li class="form-builder-rule-condition-container timeline-item" condition-index="0">
      <div class="panel panel-default">
        <div class="flex-container panel-body">
          <div class="form-group-autofit">
            <div class="form-group-item form-group-item-label form-group-item-shrink">
              <h4>
                <span class="text-truncate-inline">
                  <span class="text-truncate">if</span>
                </span>
              </h4>
            </div>
            <div class="condition-if form-group-item" condition-if-index="0">
              <div id="fields-rule">
                <p>Select Field</p>
              </div>
            </div>
            <div class="condition-operator form-group-item" condition-operator-index="0">
              <div id="fields-rule">
                <p>Select Field</p>
              </div>
            </div>
            <div class="condition-type form-group-item hide" condition-type-index="0">
              <div id="fields-rule">
                <p>Select Field</p>
              </div>
            </div>
            <div class="condition-type-value form-group-item hide" condition-type-value-index="0">
              <div id="fields-rule">
                <p>Text Field</p>
              </div>
            </div>
            <div class="condition-type-value-select form-group-item hide" condition-type-value-select-index="0">
              <div id="fields-rule">
                <p>Select Field</p>
              </div>
            </div>
          </div>
          <div class="timeline-increment">
            <span class="timeline-icon"></span>
          </div>
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`Regression LPS-85644 Should clear first operand when field is removed from pages 1`] = `
<div class="form-builder-rule-builder">
  <h2 class="form-builder-section-title text-default">rule</h2>
  <h4 class="text-default">define-condition-and-action-to-change-fields-and-elements-on-the-form</h4>
  <ul class="liferay-ddm-form-builder-rule-condition-list liferay-ddm-form-rule-builder-timeline timeline ">
    <li class="timeline-item">
      <div class="panel panel-default">
        <div class="flex-container panel-body">
          <div class="h4 panel-title">condition</div>
          <div class="btn-group dropdown" style="">
            <button class="btn btn-default dropdown-toggle dropdown-toggle-operator text-uppercase" data-toggle="dropdown" type="button" disabled="">
              <span class="dropdown-toggle-selected-value">or</span>
              <span class="caret"></span>
            </button>
            <ul class="dropdown-menu">
              <li class="logic-operator text-uppercase" data-logical-operator-value="or">
                <a href="#">or</a>
              </li>
              <li class="divider"></li>
              <li class="logic-operator text-uppercase" data-logical-operator-value="and">
                <a href="#">and</a>
              </li>
            </ul>
          </div>
          <div class="timeline-increment">
            <span class="timeline-icon"></span>
          </div>
        </div>
      </div>
    </li>
    <li class="form-builder-rule-condition-container timeline-item" condition-index="0">
      <div class="panel panel-default">
        <div class="flex-container panel-body">
          <div class="form-group-autofit">
            <div class="form-group-item form-group-item-label form-group-item-shrink">
              <h4>
                <span class="text-truncate-inline">
                  <span class="text-truncate">if</span>
                </span>
              </h4>
            </div>
            <div class="condition-if form-group-item" condition-if-index="0">
              <div id="fields-rule">
                <p>Select Field</p>
              </div>
            </div>
            <div class="condition-operator form-group-item" condition-operator-index="0">
              <div id="fields-rule">
                <p>Select Field</p>
              </div>
            </div>
            <div class="condition-type form-group-item hide" condition-type-index="0">
              <div id="fields-rule">
                <p>Select Field</p>
              </div>
            </div>
            <div class="condition-type-value form-group-item hide" condition-type-value-index="0">
              <div id="fields-rule">
                <p>Text Field</p>
              </div>
            </div>
            <div class="condition-type-value-select form-group-item hide" condition-type-value-select-index="0">
              <div id="fields-rule">
                <p>Select Field</p>
              </div>
            </div>
          </div>
          <div class="timeline-increment">
            <span class="timeline-icon"></span>
          </div>
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`RuleEditor should show all fields when fourth field is a select field 1`] = `
<div class="form-builder-rule-builder">
  <h2 class="form-builder-section-title text-default">rule</h2>
  <h4 class="text-default">define-condition-and-action-to-change-fields-and-elements-on-the-form</h4>
  <ul class="liferay-ddm-form-builder-rule-condition-list liferay-ddm-form-rule-builder-timeline timeline ">
    <li class="timeline-item">
      <div class="panel panel-default">
        <div class="flex-container panel-body">
          <div class="h4 panel-title">condition</div>
          <div class="btn-group dropdown" style="">
            <button class="btn btn-default dropdown-toggle dropdown-toggle-operator text-uppercase" data-toggle="dropdown" type="button" disabled="">
              <span class="dropdown-toggle-selected-value">or</span>
              <span class="caret"></span>
            </button>
            <ul class="dropdown-menu">
              <li class="logic-operator text-uppercase" data-logical-operator-value="or">
                <a href="#">or</a>
              </li>
              <li class="divider"></li>
              <li class="logic-operator text-uppercase" data-logical-operator-value="and">
                <a href="#">and</a>
              </li>
            </ul>
          </div>
          <div class="timeline-increment">
            <span class="timeline-icon"></span>
          </div>
        </div>
      </div>
    </li>
    <li class="form-builder-rule-condition-container timeline-item" condition-index="0">
      <div class="panel panel-default">
        <div class="flex-container panel-body">
          <div class="form-group-autofit">
            <div class="form-group-item form-group-item-label form-group-item-shrink">
              <h4>
                <span class="text-truncate-inline">
                  <span class="text-truncate">if</span>
                </span>
              </h4>
            </div>
            <div class="condition-if form-group-item" condition-if-index="0">
              <div id="fields-rule">
                <p>Select Field</p>
              </div>
            </div>
            <div class="condition-operator form-group-item" condition-operator-index="0">
              <div id="fields-rule">
                <p>Select Field</p>
              </div>
            </div>
            <div class="condition-type form-group-item " condition-type-index="0">
              <div id="fields-rule">
                <p>Select Field</p>
              </div>
            </div>
            <div class="condition-type-value form-group-item hide" condition-type-value-index="0">
              <div id="fields-rule">
                <p>Text Field</p>
              </div>
            </div>
            <div class="condition-type-value-select form-group-item " condition-type-value-select-index="0">
              <div id="fields-rule">
                <p>Select Field</p>
              </div>
            </div>
            <div class="condition-type-value-select-options form-group-item hide" condition-type-value-select-options-index="0">
              <div id="fields-rule">
                <p>Select Field</p>
              </div>
            </div>
          </div>
          <div class="timeline-increment">
            <span class="timeline-icon"></span>
          </div>
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`RuleEditor should show all fields when fourth field is a text input 1`] = `
<div class="form-builder-rule-builder">
  <h2 class="form-builder-section-title text-default">rule</h2>
  <h4 class="text-default">define-condition-and-action-to-change-fields-and-elements-on-the-form</h4>
  <ul class="liferay-ddm-form-builder-rule-condition-list liferay-ddm-form-rule-builder-timeline timeline ">
    <li class="timeline-item">
      <div class="panel panel-default">
        <div class="flex-container panel-body">
          <div class="h4 panel-title">condition</div>
          <div class="btn-group dropdown" style="">
            <button class="btn btn-default dropdown-toggle dropdown-toggle-operator text-uppercase" data-toggle="dropdown" type="button" disabled="">
              <span class="dropdown-toggle-selected-value">or</span>
              <span class="caret"></span>
            </button>
            <ul class="dropdown-menu">
              <li class="logic-operator text-uppercase" data-logical-operator-value="or">
                <a href="#">or</a>
              </li>
              <li class="divider"></li>
              <li class="logic-operator text-uppercase" data-logical-operator-value="and">
                <a href="#">and</a>
              </li>
            </ul>
          </div>
          <div class="timeline-increment">
            <span class="timeline-icon"></span>
          </div>
        </div>
      </div>
    </li>
    <li class="form-builder-rule-condition-container timeline-item" condition-index="0">
      <div class="panel panel-default">
        <div class="flex-container panel-body">
          <div class="form-group-autofit">
            <div class="form-group-item form-group-item-label form-group-item-shrink">
              <h4>
                <span class="text-truncate-inline">
                  <span class="text-truncate">if</span>
                </span>
              </h4>
            </div>
            <div class="condition-if form-group-item" condition-if-index="0">
              <div id="fields-rule">
                <p>Select Field</p>
              </div>
            </div>
            <div class="condition-operator form-group-item" condition-operator-index="0">
              <div id="fields-rule">
                <p>Select Field</p>
              </div>
            </div>
            <div class="condition-type form-group-item " condition-type-index="0">
              <div id="fields-rule">
                <p>Select Field</p>
              </div>
            </div>
            <div class="condition-type-value form-group-item " condition-type-value-index="0">
              <div id="fields-rule">
                <p>Text Field</p>
              </div>
            </div>
            <div class="condition-type-value-select form-group-item hide" condition-type-value-select-index="0">
              <div id="fields-rule">
                <p>Select Field</p>
              </div>
            </div>
            <div class="condition-type-value-select-options form-group-item hide" condition-type-value-select-options-index="0">
              <div id="fields-rule">
                <p>Select Field</p>
              </div>
            </div>
          </div>
          <div class="timeline-increment">
            <span class="timeline-icon"></span>
          </div>
        </div>
      </div>
    </li>
  </ul>
</div>
`;
